FINERACT-1707 - S3 export and dependent service

b0c1 opened the pull request:
FINERACT-1707 - S3 export and dependent service
- [x] S3 export function
- [x] list all available export-type services
- [x] refactor export service
- [x] unit test for export name generator
- [x] s3 integration test
- [x] localstack integration for testing

## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).


## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/data/ReportExportType.java:
```diff
@@ -0,0 +1,34 @@
+import lombok.Data;
+
+@Data
+public class ReportExportType implements Serializable {
```
`@RequiredArgsConstructor`

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java:
```diff
@@ -70,6 +72,29 @@ public RunreportsApiResource(final PlatformSecurityContext context, final ReadRe
+        queryParams.putAll(uriInfo.getQueryParameters());
+
+        final boolean parameterType = ApiParameterHelper.parameterType(queryParams);
+        String reportType = this.readExtraDataAndReportingService.getReportType(reportName, isSelfServiceUserReport, parameterType);
```
`this.` not needed

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/RunreportsApiResource.java:
```diff
@@ -70,6 +72,29 @@ public RunreportsApiResource(final PlatformSecurityContext context, final ReadRe
+        final boolean parameterType = ApiParameterHelper.parameterType(queryParams);
+        String reportType = this.readExtraDataAndReportingService.getReportType(reportName, isSelfServiceUserReport, parameterType);
+        ReportingProcessService reportingProcessService = this.reportingProcessServiceProvider.findReportingProcessService(reportType);
+        if (reportingProcessService == null) {
```
Let's not put this logic here but in the provider.

b0c1 replied:
After discuss, currently not changed, because used by another method.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
-    private Response exportS3(String reportName, MultivaluedMap<String, String> queryParams, Map<String, String> reportParams,
-            boolean isSelfServiceUserReport, String parameterTypeValue) {
-        throw new UnsupportedOperationException("S3 export not supported for datatables");
+    private boolean isAwsCredentialsValid() {
```
This definitely doesn't belong here. Why can't we use the AWS SDK provided `AwsCredentialsProvider`?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -18,57 +18,79 @@
+@Slf4j
 public class DatatableReportingProcessService implements ReportingProcessService {
 
+    public static final String AWS_S_3_BUCKET_NAME = "AWS_S3_BUCKET_NAME";
```
This doesn't belong here, let's use the FineractProperties for this.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+        }
+        try {
+            S3Client s3Client = S3Client.builder().region(Region.of(System.getenv(SdkSystemSetting.AWS_REGION.environmentVariable())))
+                    .credentialsProvider(EnvironmentVariableCredentialsProvider.create()).build();
```
Let's not hardcode the EnvironmentVariableCredentialsProvider, we need the `DefaultCredentialsProvider` so that it'll read from multiple places.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+        try {
+            S3Client s3Client = S3Client.builder().region(Region.of(System.getenv(SdkSystemSetting.AWS_REGION.environmentVariable())))
+                    .credentialsProvider(EnvironmentVariableCredentialsProvider.create()).build();
+            StreamingOutput output = this.readExtraDataAndReportingService.retrieveReportCSV(reportName, parameterTypeValue, reportParams,
```
`this.` not needed

galovics replied:
It's not your fault, just noting but I really really don't like this StreamingOutput interface here, it's a javax.ws.rs class which is not supposed to be on this level but on the API.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+            return Response.status(Response.Status.BAD_REQUEST).entity("AWS S3 bucket name not set").build();
+        }
+        try {
+            S3Client s3Client = S3Client.builder().region(Region.of(System.getenv(SdkSystemSetting.AWS_REGION.environmentVariable())))
```
Can't we create the S3Client upon startup?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+            output.write(byteArrayOutputStream);
+            byteArrayOutputStream.close();
+            String finalS3Path = getS3FolderName() + "" + exportFileNameGeneration(reportName, reportParams) + ".csv";
```
Just a very minor thing but can we construct this path value with String.format? Also, is it intentional to have an empty String in the middle?

Bonus: would be awesome if you'd extract this to a factory class so that you can separately write unit test for it.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+                    .credentialsProvider(EnvironmentVariableCredentialsProvider.create()).build();
+            StreamingOutput output = this.readExtraDataAndReportingService.retrieveReportCSV(reportName, parameterTypeValue, reportParams,
+                    isSelfServiceUserReport);
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
```
try with resources?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+                    RequestBody.fromBytes(byteArrayOutputStream.toByteArray()));
+            return Response.noContent().build();
+        } catch (IOException e) {
+            log.error("Error while exporting to S3", e);
```
No need for the log message if you're rethrowing the exception below.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+            return Response.noContent().build();
+        } catch (IOException e) {
+            log.error("Error while exporting to S3", e);
+            throw new IllegalStateException("Error while exporting to S3!", e);
```
Exclamation mark not necessary, is it? :)

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+        if (StringUtils.isNotBlank(folderName)) {
+            folderName = folderName + "/";
+        }
+        return folderName.trim();
```
why don't we trim beforehand? the end will not be trimmed anyway because you append a `/` at the end.
I think I'm missing more validation here since this is a user editable value (the global config). I think the configurationDomainService should take care of validating whether the folder name doesn't contain a slash at the end and whitespaces anywhere so that from a contractual point of view, whoever uses the `retrieveReportExportS3FolderName` method, they don't need to do the same check over and over again.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+    }
+
+    private String exportFileNameGeneration(String reportName, Map<String, String> reportParams) {
+        String timestamp = "_" + new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
```
Careful, this is the old Date API on one hand, it's better to use the new java.time package for this and bonus question, shall we have the business date here, the tenant datetime or the system datetime? (pointer: DateUtils class)

b0c1 replied:
It's the generation datetime, I thin it's generation timestamp

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+
+    private String exportFileNameGeneration(String reportName, Map<String, String> reportParams) {
+        String timestamp = "_" + new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
+        int reportMaximumFileName = 512 - timestamp.length() - 4;
```
Comment here please on where the 512 and 4 comes from.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/DatatableReportingProcessService.java:
```diff
@@ -90,13 +112,64 @@ private Response exportJSON(String reportName, MultivaluedMap<String, String> qu
+        String timestamp = "_" + new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
+        int reportMaximumFileName = 512 - timestamp.length() - 4;
+        String fileName = reportName;
+        if (reportParams != null) {
```
This should really be extracted into another class & method to be able to properly unit test it.

b0c1 committed:
FINERACT-1707 - S3 export and dependent service
- [x] S3 export function
- [x] list all available export type services
- [x] refactor export service
- [x] unit test for export name generator
- [x] s3 integration test
- [x] localstack integration for testing

vidakovic commented:
@galovics @b0c1 ... is there any reason why we use the database (global configuration service) to configure these things? Why wouldn't we put this in application.properties?

galovics commented on .github/workflows/build-mariadb.yml:
```diff
@@ -52,12 +52,39 @@ jobs:
+          pip install --upgrade pyopenssl
+          echo "Install localstack"
+          pip install localstack awscli-local[ver1] # install LocalStack cli and awslocal
+          docker pull localstack/localstack         # Make sure to pull the latest version of the image
```
Aren't we mixing up 2 different approaches here? One is starting localstack via its own CLI tool and one via docker?

b0c1 replied:
It's come from localstack official documentation. It's using the docker but doesn't pull the image.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/export/CsvDatatableReportExportServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+    private final ReadReportingService readExtraDataAndReportingService;
+
+    @Override
+    public Response export(String reportName, MultivaluedMap<String, String> queryParams, Map<String, String> reportParams,
```
I'm not gonna push back on the PR because of this but just a note. This is a service layer class yet it's returning a Response which is a view layer class. We're mixing up the layers.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/export/CsvDatatableReportExportServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+                isSelfServiceUserReport);
+
+        return Response.ok().entity(result).type("text/csv")
+                .header("Content-Disposition", "attachment;filename=" + reportName.replaceAll(" ", "") + ".csv").build();
```
When it comes to the attachment's name, shouldn't we do the same thing as for DatatableExportUtil#normalizeFolderName?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/export/S3DatatableReportExportServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+    private final ConfigurationDomainService configurationDomainService;
+    private final S3Client s3Client;
+
+    @Value("${fineract.report.export.s3.bucket:reports}")
```
FineractProperties class?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/export/S3DatatableReportExportServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+            try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream()) {
+                output.write(byteArrayOutputStream);
+                String folder = configurationDomainService.retrieveReportExportS3FolderName();
+                String filePath = DatatableExportUtil.generateDatatableExportFileName(AWS_S3_MAXIMUM_KEY_LENGTH, folder, ".csv", reportName,
```
Can't we change the extension parameter to not need the dot prefix?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,75 @@
+        return DefaultAwsRegionProviderChain.builder().build();
+    }
+
+    @Bean("s3Client")
```
No need to specify the bean name, it'll automatically get the method name.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,75 @@
+    @Conditional(AmazonS3ConfigCondition.class)
+    public S3Client s3ClientTest(DefaultCredentialsProvider awsCredentialsProvider, AwsRegionProvider awsRegionProvider) {
+        S3ClientBuilder builder = S3Client.builder();
+        String env = System.getenv().getOrDefault("AWS_ENDPOINT_URL", "");
```
You can use the Environment class from Spring.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Bean
+    @ConditionalOnBean(S3Client.class)
+    public S3DatatableReportExportServiceImpl s3DatatableReportExportServiceImpl(ReadReportingService reportServiceImpl,
```
I don't think this bean definition should be here. Btw why don't we remove the definition completely and use the Component annotation on the class itelf?

b0c1 replied:
It should. Initialization moved to a different Configuration but `S3Client` must be available (or not) in the BeanFactory, because of this `@Component` will not work with `@ConditionOnBean` annotation

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,75 @@
+        return S3Client.builder().credentialsProvider(awsCredentialsProvider).region(awsRegionProvider.getRegion()).build();
+    }
+
+    @Bean("s3Client")
```
Btw I'd approach this differently a bit. I'm not saying it's not going to work but would be cleaner.

Let's assume we have this config:
```
    @Bean
    @Conditional(AmazonS3ConfigCondition.class)
    public S3Client s3Client(DefaultCredentialsProvider awsCredentialsProvider, AwsRegionProvider awsRegionProvider) {
        return S3Client.builder().credentialsProvider(awsCredentialsProvider).region(awsRegionProvider.getRegion()).build();
    }
```
Let's create a custom interface:
```
interface S3ClientCustomizer {
    void customize(S3ClientBuilder builder);
}
```
Then let's change the bean config:
```
    @Bean
    @Conditional(AmazonS3ConfigCondition.class)
    public S3Client s3Client(DefaultCredentialsProvider awsCredentialsProvider, AwsRegionProvider awsRegionProvider, List<S3ClientCustomizer> customizers) {
 
 
 
 S3ClientBuilder builder = S3Client.builder().credentialsProvider(awsCredentialsProvider).region(awsRegionProvider.getRegion());
        customizers.forEach(c -> c.customize(builder));
        return builder.build();
    }
```
And then we create a customizer for the tests:
```
@Component
@RequiredArgsConstructor
@Profile("test")
public class LocalstackS3ClientCustomizer implements S3ClientCustomizer {
    private final Environment environment;

    @Override
    public void customize(S3ClientBuilder builder) {
        String env = environment.....
        if (StringUtil.isNotBlank(env)) {
            builder.endpointOverride(URI.create(env)).forcePathStyle(true);
        }
    }
}
```

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3ConfigCondition.java:
```diff
@@ -0,0 +1,53 @@
+
+public class AmazonS3ConfigCondition implements Condition {
+
+    @Value("${fineract.report.export.s3.enabled}")
```
FineractProperties?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3ConfigCondition.java:
```diff
@@ -0,0 +1,53 @@
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.regions.providers.AwsRegionProvider;
+
+public class AmazonS3ConfigCondition implements Condition {
```
You can use PropertiesCondition.

b0c1 replied:
It provides a complex condition to enable S3 configuration

galovics replied:
The point of using PropertiesCondition is that it's gonna be a compilation error if anything changes while in case you rely on the property strings, you don't get any compilation-time checks.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3ConfigCondition.java:
```diff
@@ -0,0 +1,53 @@
+
+    private boolean isAwsCredentialValid(ConditionContext context) {
+        try {
+            context.getBeanFactory().getBean(AwsCredentialsProvider.class).resolveCredentials();
```
This is not necessarily the best idea. You're accessing the container before it even gets properly initialized. 
It's better if you explicitly instantiate the default credentials provider and region provider to test this.

https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/annotation/Condition.html

> Conditions must follow the same restrictions as [BeanFactoryPostProcessor](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/config/BeanFactoryPostProcessor.html) and take care to never interact with bean instances.

galovics commented on integration-tests/src/test/java/org/apache/fineract/integrationtests/client/ReportExportTest.java:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @author Michael Vorburger.ch
+ */
+@Tag("github_only")
```
Do we need this tag?

galovics commented on integration-tests/src/test/java/org/apache/fineract/integrationtests/client/ReportExportTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Test
+    @GithubOnly
+    @Disabled
```
Can't we put the `@Disabled` annotation into the GitHubOnly annotation?
This is kinda confusing to me that it's disabled but it's actually not. Also, if you move this to the GitHubOnly annotation, let's put a comment there and explain why.

galovics commented on integration-tests/src/test/java/org/apache/fineract/integrationtests/GithubOnly.java:
```diff
@@ -0,0 +1,30 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@EnabledIfEnvironmentVariable(named = "GITHUB_ACTIONS", matches = "true")
+public @interface GithubOnly {}
```
Let's rename this to CIOnly.

galovics commented:
> @galovics @b0c1 ... is there any reason why we use the database (global configuration service) to configure these things? Why wouldn't we put this in application.properties?

@vidakovic we wanna be able to change the configuration value runtime.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3Config.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Bean
+    @ConditionalOnBean(S3Client.class)
+    public S3DatatableReportExportServiceImpl s3DatatableReportExportServiceImpl(ReadReportingService reportServiceImpl,
```
I think you forgot to remove this bean since it's defined in AmazonS3ExportConfiguration.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/LocalstackS3ClientCustomizer.java:
```diff
@@ -0,0 +1,43 @@
+@Component
+@RequiredArgsConstructor
+@Profile("test")
+public class LocalstackS3ClientCustomizer implements S3ClientCustomizer {
```
Splendid ;)

b0c1 replied:
❤️

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3ConfigCondition.java:
```diff
@@ -0,0 +1,51 @@
+    public boolean matches(ConditionContext context, AnnotatedTypeMetadata metadata) {
+        Boolean s3ReportEnabled = context.getEnvironment().getRequiredProperty("fineract.report.export.s3.enabled", Boolean.class);
+        boolean s3BucketIsSet = StringUtils.isNotBlank(context.getEnvironment().getProperty("fineract.report.export.s3.bucket", ""));
+        log.info("AWS S3 Enabled: {}", s3ReportEnabled);
```
Let's either remove this log message or also log out that the credentials are valid. I think this'll send the wrong message that S3 is enabled even in case the credentials are invalid.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/s3/AmazonS3ConfigCondition.java:
```diff
@@ -0,0 +1,51 @@
+            DefaultAwsRegionProviderChain.builder().build().getRegion();
+            return true;
+        } catch (Exception e) {
+            log.warn("AWS credential is not valid", e);
```
So I think this is not necessarily the best idea either since:
1. the logic is clear, in case its invalid, we'll return false
2. we're saying the credentials are invalid yet in case of incorrectly set region, we also return false and claim that the creds were invalid.

vidakovic commented:
> > @galovics @b0c1 ... is there any reason why we use the database (global configuration service) to configure these things? Why wouldn't we put this in application.properties?
> 
> @vidakovic we wanna be able to change the configuration value runtime.

... that is a solved problem in Spring/Boot (search: "reloadable configuration")... I'd say not a strong argument to bury it in the database...

galovics commented:
> > > @galovics @b0c1 ... is there any reason why we use the database (global configuration service) to configure these things? Why wouldn't we put this in application.properties?
> > 
> > 
> > @vidakovic we wanna be able to change the configuration value runtime.
> 
> ... that is a solved problem in Spring/Boot (search: "reloadable configuration")... I'd say not a strong argument to bury it in the database...

I'm not saying you're wrong but if I remember correctly reloading configuration is only supported with Spring Cloud and done by annotating classes with `@RefreshScope`. Now, upon refresh, you'll have the choice whether to just "refresh" the application context, reload or shutdown-reload it. 

I've used it with a cloud project but I think fineract in its current form is not suitable for this and could cause issues with context refreshes.

**TL;DR;** I think going with refreshable properties with Spring's approach is way too premature for Fineract.
